framework/Steps.py

```diff
@@ -679,82 +679,130 @@ def _localTakeAstepRun(self,inDictionary):
+                self.raiseAWarning('The job "'+finishedJob.identifier+'" has been submitted '+ str(self.failureHandling['repetitions'])+' times, failing all the times!!!')
+            if sampler.ableToHandelFailedRuns:
+              self.raiseAWarning('The sampler/optimizer "'+sampler.type+'" is able to handle failed runs!')
+            #pop the failed job from the list
+            finishedJobList.pop(finishedJobList.index(finishedJob))
+        if sampler.batch > 0: # TODO: should be consistent, if no batching should batch size be 1 or 0 ?
+          # if sampler claims it's batching, then only collect once, since it will collect the batch
+          # together, not one-at-a-time
+          sampler.finalizeActualSampling(finishedJob,model,inputs)
+        else:
+          # sampler isn't intending to batch, so we send them in one-at-a-time as per normal
+          for finishedJob in finishedJobList:
+            # finalize actual sampler
+            sampler.finalizeActualSampling(finishedJob,model,inputs)
+        for finishedJob in finishedJobList:
+          finishedJob.trackTime('step_finished')
 
         # terminate jobs as requested by the sampler, in case they're not needed anymore
         ## TODO is this a safe place to put this?
         ## If it's placed after adding new jobs and IDs are re-used i.e. for failed tests,
         ## -> then the new jobs will be killed if this is placed after new job submission!
         jobHandler.terminateJobs(sampler.getJobsToEnd(clear=True))
 
-        # add new jobs
-        isEnsemble = isinstance(model, Models.EnsembleModel)
+        # add new jobs, for DET-type samplers
         # put back this loop (do not take it away again. it is NEEDED for NOT-POINT samplers(aka DET)). Andrea
-        ## In order to ensure that the queue does not grow too large, we will
-        ## employ a threshold on the number of jobs the jobHandler can take,
-        ## in addition, we cannot provide more jobs than the sampler can provide.
-        ## So, we take the minimum of these two values.
-        self.raiseADebug('Testing if the sampler is ready to generate a new input')
-        for _ in range(min(jobHandler.availability(isEnsemble), sampler.endJobRunnable())):
-
-          if sampler.amIreadyToProvideAnInput():
-            try:
-              newInput = self._findANewInputToRun(sampler, model, inputs, outputs, jobHandler)
-              if newInput is not None:
-                model.submit(newInput, inDictionary[self.samplerType].type, jobHandler, **copy.deepcopy(sampler.inputInfo))
-            except utils.NoMoreSamplesNeeded:
-              self.raiseAMessage(' ... Sampler returned "NoMoreSamplesNeeded".  Continuing...')
-              break
-          else:
-            self.raiseADebug(' ... sampler has no new inputs currently.')
-            break
+        # NOTE for non-DET samplers, this check also happens outside this collection loop
+        if sampler.onlySampleAfterCollecting:
```

Comment:
onlySampleAfterCollecting, this flag need to be revisited, and updated

Reply:
An issue need to be created if not addressed by this PR

Reply:
see issue #1402